### PR TITLE
Handle new volume source types on web console

### PR DESCRIPTION
**Pull request by fabianofranz**
Adding specific details about `DownwardAPIVolumeSource` and `ConfigMapVolumeSource` volume source types to the web console. 

Fixes https://github.com/openshift/origin/issues/7885.

@jwforres PTAL

### CODE REVIEW

File: assets/app/views/_volumes.html [hunk 1]
```diff
@@ -46,5 +46,23 @@
...
+        <dd>{{item.path}}&#8201;&#8594;&#8201;pod field "{{item.fieldRef.fieldPath}}"</dd>
+      </div>
+    </div>
+    <div ng-if="volume.configMap">
+      <dt>Type:</dt>
+      <dd>config map</dd>
+      <dt>Referenced config map:</dt>
+      <dd>{{volume.configMap.name}}</dd>
+      <div ng-repeat="item in volume.configMap.items">        
+        <dt>Projected key to path:</dt>
```
Comment (jwforres) [inline assets/app/views/_volumes.html]: hmm is this what the CLI shows?

Comment (jwforres) [inline assets/app/views/_volumes.html]: also config map and secret should be really similar in theory so if we need to update the one for secret we should do that

Comment (fabianofranz) [inline assets/app/views/_volumes.html]: What about "Volume file" instead of "Projected key to path", like in downward api?

The CLI doesn't show items, only the type and the referenced config map (called simply "Name").

**Comment by fabianofranz** [comment]
Also I added a `small text-muted` description to every "type", using the same info in the CLI describer.

File: assets/app/views/_volumes.html [hunk 2]
```diff
@@ -8,13 +8,19 @@
...
+      </dd>
       <dt>Secret name:</dt>
       <dd>{{volume.secret.secretName}}</dd>
     </div>
     <div ng-if="volume.persistentVolumeClaim">
       <dt>Type:</dt>
-      <dd>persistent volume claim</dd>
+      <dd>
+				persistent volume claim
+        <span class="small text-muted">(a reference to a PersistentVolumeClaim in the same namespace)</span>
```
Comment (jwforres) [inline assets/app/views/_volumes.html]: a bit wordy, maybe just "a reference to a Persistent Volume Claim"

File: assets/app/views/_volumes.html [hunk 3]
```diff
@@ -8,13 +8,19 @@
   <dl class="dl-horizontal left">
     <div ng-if="volume.secret">
       <dt>Type:</dt>
-      <dd>secret</dd>
+      <dd>
+        secret
+        <span class="small text-muted">(a volume populated by a Secret)</span>
```
Comment (jwforres) [inline assets/app/views/_volumes.html]: feels redundant to repeat the word volume here, "populated by a Secret at container creation" maybe?

File: assets/app/views/_volumes.html [hunk 4]
```diff
@@ -25,13 +31,19 @@
...
+       </dd>
       <dt>Path:</dt>
       <dd>{{volume.hostPath.path}}</dd>
     </div>
     <div ng-if="volume.emptyDir">
       <dt>Type:</dt>
-      <dd>empty dir</dd>
+      <dd>
+        empty dir
+       <span class="small text-muted">(a temporary directory that shares a pod's lifetime)</span>
```
Comment (jwforres) [inline assets/app/views/_volumes.html]: "temporary directory destroyed with the pod" ?   drop the leading "a"

File: assets/app/views/_volumes.html (hunk 2 above)
Comment (jwforres) [inline assets/app/views/_volumes.html]: actually drop the leading "a"

File: assets/app/views/_volumes.html (hunk 3 above)
Comment (jwforres) [inline assets/app/views/_volumes.html]: sorry to be consistent with your later ones "populated by a Secret when the pod is created"

File: assets/app/views/_volumes.html [hunk 5]
```diff
@@ -40,11 +52,38 @@
     </div>
     <div ng-if="volume.gitRepo">
       <dt>Type:</dt>
-      <dd>git repo</dd>
+      <dd>
+        git repo
+       <span class="small text-muted">(a volume that is pulled from git when the pod is created)</span>
```
Comment (jwforres) [inline assets/app/views/_volumes.html]: "pulled from git when the pod is created"

File: assets/app/views/_volumes.html [hunk 6]
```diff
@@ -40,11 +52,38 @@
...
       <dt>Repository:</dt>
       <dd>{{volume.gitRepo.repository}}</dd>
       <dt ng-if-start="volume.gitRepo.revision">Revision:</dt>
       <dd ng-if-end>{{volume.gitRepo.revision}}</dd>
     </div>
+    <div ng-if="volume.downwardAPI">
+      <dt>Type:</dt>
+      <dd>
+        downward API
+        <span class="small text-muted">(a volume populated by information about the pod)</span>
```
Comment (jwforres) [inline assets/app/views/_volumes.html]: "populated by information about the pod"

Comment (jwforres) [inline assets/app/views/_volumes.html]: sorry "populated with information about the pod"

File: assets/app/views/_volumes.html [hunk 7]
```diff
@@ -40,11 +52,38 @@
...
+      <div ng-repeat="item in volume.downwardAPI.items">
+        <dt>Volume file:</dt>
+        <dd>{{item.fieldRef.fieldPath}}&#8201;&#8594;&#8201;{{item.path}}</dd>
+      </div>
+    </div>
+    <div ng-if="volume.configMap">
+      <dt>Type:</dt>
+      <dd>
+        config map
+        <span class="small text-muted">(a volume populated by a config map)</span>
```
Comment (jwforres) [inline assets/app/views/_volumes.html]: "populated by a Config Map"  if you capitalized the other kinds we should be consistent with this one

File: assets/app/views/_volumes.html (hunk 2 above)
Comment (fabianofranz) [inline assets/app/views/_volumes.html]: Sounds good, fixed.

File: assets/app/views/_volumes.html [hunk 8]
```diff
@@ -40,11 +52,38 @@
...
+       </dd>
+      <dt>Name:</dt>
+      <dd>{{volume.configMap.name}}</dd>
+      <div ng-repeat="item in volume.configMap.items">
+        <dt>Volume file:</dt>
+        <dd>{{item.path}}&#8201;&#8594;&#8201;{{item.key}}</dd>
```
Comment (jwforres) [inline assets/app/views/_volumes.html]: hmm the problem with this is that I don't necessarily understand that the value being shown for item.key is that key of the config map, maybe

**Key to File:** foo &#8201;&#8594;&#8201; /some/path

or something like that

Comment (jwforres) [inline assets/app/views/_volumes.html]: @spadgett what do you think about this?  for config map and for secret.  can't decide

**Comment by fabianofranz** [comment]
@jwforres comments up to this point addressed.

**Comment by fabianofranz** [comment]
Worth to mention that the CLI describer has more volume source types: GCEPersistentDisk, AWSElasticBlockStore, NFS, ISCSI, Glusterfs and RBD. Should we add them too?

File: assets/app/views/_volumes.html (hunk 8 above)
Comment (fabianofranz) [inline assets/app/views/_volumes.html]: Here's how they look like in the CLI, for comparison:

```
Volumes:
  default-token-x0svm:
    Type:	Secret (a volume populated by a Secret)
    SecretName:	default-token-x0svm
  localvol1:
    Type:	ConfigMap (a volume populated by a ConfigMap)
    Name:	pod/pod-volume-downward-api
  localvol2:
    Type:	DownwardAPI (a volume populated by information about the pod)
    Items:
      metadata.name -> path/to/file

```

**Comment by fabianofranz** [comment]
Related CLI code [here](https://github.com/openshift/origin/blob/master/Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubectl/describe.go#L658-L797).

**Comment by jwforres** [comment]
No not for now, we've always said we weren't going to go crazy trying to
show all the different volume types from the console and just hit the most
important ones.  Maybe we can get someone from the storage UI team to go do
that at some point in 1.3 or 1.4.

File: assets/app/views/_volumes.html (hunk 8 above)
Comment (jwforres) [inline assets/app/views/_volumes.html]: heh, yeah... telling you your volume is a volume is a bit redundant, I
assume that describer is coming from k8s?  cc @smarterclayton

so ConfigMap and Secret in the CLI don't even list out the keys to paths?

Comment (smarterclayton) [inline assets/app/views/_volumes.html]: No idea.

Comment (fabianofranz) [inline assets/app/views/_volumes.html]: Yep, they come [from Kube](https://github.com/openshift/origin/blob/master/Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubectl/describe.go#L658-L797). Only DownwardAPI lists the items.

**Comment by jwforres** [comment]
@fabianofranz what does `oc volumes dc/foo --list` show for those?

**Comment by fabianofranz** [comment]
@jwforres `volumes --list` seems more focused on how it's mounted inside the containter, instead of the actual volume specs:

```
pods/pod-volume-config-map
  localvol
    mounted at /usr/share/nginx/html/test
  secret/default-token-x0svm as default-token-x0svm
    mounted at /var/run/secrets/kubernetes.io/serviceaccount
```

**Comment by jwforres** [comment]
ok, i think this is good for now then [merge]
